settings/src/test/java/org/odk/collect/settings/ODKAppSettingsMigratorTest.java

```diff
@@ -322,31 +322,52 @@ public void when_AllowOtherWaysOfEditingFormIsDisabled_thenSaveAsDraftShouldBeDi
         initSettings(protectedSettings,
                 ProtectedProjectKeys.ALLOW_OTHER_WAYS_OF_EDITING_FORM, false,
                 ProtectedProjectKeys.KEY_SAVE_AS_DRAFT, true,
-                ProtectedProjectKeys.KEY_FINALIZE, false
+                ProtectedProjectKeys.KEY_FINALIZE_IN_FORM_ENTRY, false
         );
 
         runMigrations();
 
         assertThat(protectedSettings.contains(ProtectedProjectKeys.ALLOW_OTHER_WAYS_OF_EDITING_FORM), equalTo(true));
         assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.ALLOW_OTHER_WAYS_OF_EDITING_FORM), equalTo(false));
         assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.KEY_SAVE_AS_DRAFT), equalTo(false));
-        assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.KEY_FINALIZE), equalTo(true));
+        assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.KEY_FINALIZE_IN_FORM_ENTRY), equalTo(true));
     }
 
     @Test
     public void when_AllowOtherWaysOfEditingFormIsEnabled_thenDoNotUpdateSaveAsDraftOrFinalize() {
         initSettings(protectedSettings,
                 ProtectedProjectKeys.ALLOW_OTHER_WAYS_OF_EDITING_FORM, true,
                 ProtectedProjectKeys.KEY_SAVE_AS_DRAFT, true,
-                ProtectedProjectKeys.KEY_FINALIZE, false
+                ProtectedProjectKeys.KEY_FINALIZE_IN_FORM_ENTRY, false
         );
 
         runMigrations();
 
         assertThat(protectedSettings.contains(ProtectedProjectKeys.ALLOW_OTHER_WAYS_OF_EDITING_FORM), equalTo(true));
         assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.ALLOW_OTHER_WAYS_OF_EDITING_FORM), equalTo(true));
         assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.KEY_SAVE_AS_DRAFT), equalTo(true));
-        assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.KEY_FINALIZE), equalTo(false));
+        assertThat(protectedSettings.getBoolean(ProtectedProjectKeys.KEY_FINALIZE_IN_FORM_ENTRY), equalTo(false));
+    }
+
+    @Test
+    public void migratesFinalizeInFormEntryToNewKey() {
```

Comment:
I couldn't think of another way to support upgrades and imports from up to date versions other than creating a new key. Otherwise, we'd end up with QR codes from up to date versions with only "finalise in form entry" disabled also disabling bulk finalise.

Reply:
This means that if you have the existing `finalize` key in your QR code, it will forever more turn off both finalization at end of form and bulk finalize when you scan it in, right? That sounds fine and good.

If you start from a clean Collect install and you hide finalization from form entry, that will no affect bulk finalization. If you build a QR code from those settings, bulk finalization will be enabled unless you also explicitly unchecked it.

Unless I made a mistake in my interpretation, I think the approach is great.